pkg/trace/span.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trace
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+type corrIDKey string
+
+// CorrID is a correlation ID that the cluster API provider
+// sends with all API requests to Azure. Do not create one
+// of these manually. Instead, use the CtxWithCorrelationID function
+// to create one of these within a context.Context.
+type CorrID string
+
+const corrIDKeyVal corrIDKey = "x-ms-correlation-id"
+
+// StartSpan starts a new span with the given context and tracer, using
+// the given spanName for the span's name. This function also ensures
+// that a new correlation ID is set on the passed context if none
+// existed already. Callers should make sure to do two things with
+// the return values of this function. First, only use the new context
+// returned from this function. Do not use the one you passed. The
+// new one will represent the new span and contain the new correlation
+// ID (if one was created and set). Second, make sure to close the
+// span when you're done with it. Usually this is done with a defer.
+// See below for an example:
+//
+//	ctx := context.Background()
+//	ctx, corrID, newSpan, err := StartSpan(ctx, tele.Tracer(), "myNewSpan")
+//	if err != nil {
+//		log.Fatal(err)
+//	}
+//	defer newSpan.End()
+//	doSomething(ctx)
+func StartSpan(
```

Comment:
WDYT of something like this:

In `./util/tele.go`, we wrap the `otel.Tracer` with our own struct that implements `trace.Tracer`. We then write our own `Start(...)` func which would then delegate back to `otel.Tracer().Start(...)`. That way the interface can stay the same.
```go
package tele

type tracer struct {
    otel.Tracer
}

func (t *tracer) Start(ctx, opName string, opt ...otel.SpanOption) (context.Context, otel.Span) {
    // add correlationID to ctx and continue as normal
    // return the inner implementation of Start
    return t.Tracer.Start(ctx, opName, opt...)
}

```

Then either in the same package or elsewhere, add context helpers to set / fetch the correlationID.

```go
package corr  // or something else...

// FromCtx or some better name you think of
func FromCtx(ctx context.Context) (ctx, string) {
    // create new context.WithValue containing correlationID and return both if it does not exist, else return context and correlationID
}

```

Reply:
Sure, happy to do that! A few questions before I get started, just to make sure I'm on the same page

- I'm assuming that you'd want to still call `tele.Tracer().Start(...)` in all the reconcilers, correct?
- Are you ok with creating our own tracer instance (the wrapper you mentioned) and then passing that -- instead of the default one -- to `otel.SetTracerProvider` in `main.go`?
- Do we need `FromCtx` to be exported at all? It seems like it would only be used inside the same package as the custom tracer, but not sure if you're thinking something more

Reply:
> I'm assuming that you'd want to still call tele.Tracer().Start(...) in all the reconcilers, correct?

If possible, that would be great.

> Are you ok with creating our own tracer instance (the wrapper you mentioned) and then passing that -- instead of the default one -- to otel.SetTracerProvider in main.go?

:100: 

> Do we need FromCtx to be exported at all? It seems like it would only be used inside the same package as the custom tracer, but not sure if you're thinking something more

I was adding that because I saw you were returning the correlationID in your `StartSpan` func. This led me to think you wanted a way to access it from the context. I think one might need to access the correlationID from the context to be able to add that as a HTTP header for the Azure SDK for Go clients.

Reply:
perfect - I'll make these changes.

regarding `FromCtx` - just in case someone wants to access the correlation ID, I'll make `FromCtx` exported like you said.

Reply:
@devigned just an update, so you know I haven't forgotten about this! I'm looking through the implementation & docs trying to figure out a way to create a new [`TracerProvider`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#TracerProvider) that we can pass to the `SetTracerProvider` function in `main.go`. That would need to create a new `Tracer` which wraps the implementation you referenced above. I think that's going to require a lot of plumbing, which we could make unnecessary if we called a custom function to get either:

- get a new tracer that which wraps the underlying tracer implementation, like what you showed here
- create a new span, which is what's already in this PR

I'll keep digging, but let me know about going with either of the above options in case there isn't a good way to wrap everything we need

Reply:
Another option would be to create a custom `TracerProvider` that wraps the jaeger one. Let me know your thoughts on that too?

Reply:
Sorry for the delay, @arschles. I was thinking something more like this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/master...devigned:corr

The above diff should add the correlationID to the context without needing to manipulate the signature of the Tracer. wdyt?

Reply:
Hi @devigned I see what you mean now! Thanks, I'll implement that.